File: python/llmix/key_pool.py
"""Key rotation pool for API key management.

Round-robin key selection with dead-key tracking.
On 429: rotate() advances to next key.
On 401/403: mark_dead() permanently removes a key from rotation.
Thread-safe via threading.Lock.

Env var convention:
  {PROVIDER}_KEYS  — comma-separated list (preferred)
  {PROVIDER}_API_KEY — single key fallback (pool of 1)
"""

import os
import threading


class KeyPoolExhaustedError(Exception):
    """All keys in the pool are dead (401/403)."""


class KeyPool:
    """Round-robin pool of API keys with dead-key tracking."""

    def __init__(self, keys: list[str]) -> None:
        cleaned = [k.strip() for k in keys if k.strip()]
        if not cleaned:
            raise ValueError("KeyPool requires at least one non-empty key")
        self._keys = list(dict.fromkeys(cleaned))
        self._dead: set[str] = set()
        self._idx = 0
        self._lock = threading.Lock()

    def select(self) -> str:
        """Return the current key, skipping dead keys.

        Raises KeyPoolExhaustedError if all keys are dead.
        """
        with self._lock:
            if len(self._dead) >= len(self._keys):
                raise KeyPoolExhaustedError(f"All {len(self._keys)} keys are dead (401/403). Replace keys or wait for provider resolution.")

            for offset in range(len(self._keys)):
                idx = (self._idx + offset) % len(self._keys)
                key = self._keys[idx]
                if key not in self._dead:
                    self._idx = idx
                    return key

            raise KeyPoolExhaustedError("All keys are dead")

    def rotate(self) -> None:
        """Advance to the next key (called on 429)."""
        with self._lock:
            if len(self._dead) >= len(self._keys):
                return

            for offset in range(1, len(self._keys) + 1):
                idx = (self._idx + offset) % len(self._keys)
                if self._keys[idx] not in self._dead:
                    self._idx = idx
                    return

    def mark_dead(self, key: str) -> None:
        """Mark a key as permanently failed (called on 401/403)."""
        with self._lock:
            if key not in self._keys:
                raise ValueError("Key not in pool: cannot mark unknown key as dead")
            self._dead.add(key)

    def is_exhausted(self) -> bool:
        """Return True if all keys are dead."""
        with self._lock:
            return len(self._dead) >= len(self._keys)

    @property
    def alive_count(self) -> int:
        """Number of keys still alive."""
        with self._lock:
            return len(self._keys) - len(self._dead)

    @property
    def total_count(self) -> int:
        """Total number of keys in the pool."""
        return len(self._keys)


def load_keys_from_env(provider: str) -> KeyPool:
    """Create a KeyPool from environment variables.

    Checks {PROVIDER}_KEYS first (comma-separated), then falls back
    to {PROVIDER}_API_KEY (pool of 1).

    Args:
        provider: Provider name (e.g. "OPENAI", "ANTHROPIC").
                  Converted to uppercase automatically.

    Raises:
        ValueError: If neither env var is set.
    """
    provider_upper = provider.upper()

    # Try {PROVIDER}_KEYS first (comma-separated pool)
    keys_var = f"{provider_upper}_KEYS"
    keys_raw = os.getenv(keys_var, "")
    if keys_raw.strip():
        return KeyPool(keys_raw.split(","))

    # Fall back to {PROVIDER}_API_KEY (single key)
    key_var = f"{provider_upper}_API_KEY"
    single_key = os.getenv(key_var, "")
    if single_key.strip():
        return KeyPool([single_key])

    raise ValueError(f"No API keys found for {provider_upper}. Set {keys_var} (comma-separated) or {key_var}.")

File: packages/llmix/typescript/src/key-pool.ts
/**
 * Key rotation pool for API key management.
 *
 * Round-robin key selection with dead-key tracking.
 * On 429: rotate() advances to next key.
 * On 401/403: markDead() permanently removes a key from rotation.
 * Single-threaded JS — no lock needed, just atomic index.
 *
 * Env var convention:
 *   {PROVIDER}_KEYS  — comma-separated list (preferred)
 *   {PROVIDER}_API_KEY — single key fallback (pool of 1)
 */

export class KeyPoolExhaustedError extends Error {
  constructor(message: string) {
    super(message);
    this.name = "KeyPoolExhaustedError";
  }
}

export class KeyPool {
  private readonly keys: string[];
  private readonly dead: Set<string> = new Set();
  private idx: number = 0;

  constructor(keys: string[]) {
    const cleaned = [...new Set(keys.map((k) => k.trim()).filter((k) => k.length > 0))];
    if (cleaned.length === 0) {
      throw new Error("KeyPool requires at least one non-empty key");
    }
    this.keys = cleaned;
  }

  /** Return the current key, skipping dead keys. */
  select(): string {
    if (this.isExhausted()) {
      throw new KeyPoolExhaustedError(
        `All ${this.keys.length} keys are dead (401/403). ` +
          "Replace keys or wait for provider resolution.",
      );
    }
    for (let offset = 0; offset < this.keys.length; offset++) {
      const idx = (this.idx + offset) % this.keys.length;
      const key = this.keys[idx] as string;
      if (!this.dead.has(key)) {
        this.idx = idx;
        return key;
      }
    }
    throw new KeyPoolExhaustedError("All keys are dead");
  }

  /** Advance to the next key (called on 429). */
  rotate(): void {
    if (this.isExhausted()) {
      return;
    }
    for (let offset = 1; offset <= this.keys.length; offset++) {
      const idx = (this.idx + offset) % this.keys.length;
      if (!this.dead.has(this.keys[idx] as string)) {
        this.idx = idx;
        return;
      }
    }
  }

  /** Mark a key as permanently failed (called on 401/403). */
  markDead(key: string): void {
    if (this.keys.includes(key)) {
      this.dead.add(key);
    }
  }

  /** Return true if all keys are dead. */
  isExhausted(): boolean {
    return this.dead.size >= this.keys.length;
  }

  /** Number of keys still alive. */
  get aliveCount(): number {
    return this.keys.length - this.dead.size;
  }

  /** Total number of keys in the pool. */
  get totalCount(): number {
    return this.keys.length;
  }
}

/**
 * Create a KeyPool from environment variables.
 *
 * Checks {PROVIDER}_KEYS first (comma-separated), then falls back
 * to {PROVIDER}_API_KEY (pool of 1).
 *
 * @param provider - Provider name (e.g. "OPENAI", "ANTHROPIC"). Converted to uppercase.
 * @throws Error if neither env var is set.
 */
export function loadKeysFromEnv(provider: string): KeyPool {
  // Normalize non-identifier characters (e.g. '-' in "sno-gpu") to '_' so
  // env var names are valid POSIX/Kubernetes identifiers (SNO_GPU_KEYS,
  // SNO_GPU_API_KEY).
  const providerUpper = provider.toUpperCase().replace(/-/g, "_");

  // Try {PROVIDER}_KEYS first (comma-separated pool)
  const keysVar = `${providerUpper}_KEYS`;
  const keysRaw = process.env[keysVar] ?? "";
  if (keysRaw.trim()) {
    return new KeyPool(keysRaw.split(","));
  }

  // Fall back to {PROVIDER}_API_KEY (single key)
  const keyVar = `${providerUpper}_API_KEY`;
  const singleKey = process.env[keyVar] ?? "";
  if (singleKey.trim()) {
    return new KeyPool([singleKey]);
  }

  throw new Error(
    `No API keys found for ${providerUpper}. Set ${keysVar} (comma-separated) or ${keyVar}.`,
  );
}
